checker/src/main/java/org/checkerframework/checker/regex/RegexAnnotatedTypeFactory.java

```diff
@@ -487,17 +487,20 @@ public static int getGroupCount(@Regex String regexp) {
         // TODO check and ask for missing cases or alternate ways.
         List<Integer> nonNullGroups = new ArrayList<>();
         int n = getGroupCount(regexp);
-        for (int i = 0; i <= n; i++) {
+        for (int i = 1; i <= n; i++) {
             nonNullGroups.add(i);
         }
         ArrayDeque<Integer> openingIndices = new ArrayDeque<>();
         int group = 0;
         boolean squareBracketOpen = false;
         boolean escaped = false;
         boolean notCapturing = false;
+        boolean quoted = false;
         for (int i = 0; i < regexp.length(); i++) {
-            if (!escaped && !squareBracketOpen && regexp.charAt(i) == '(') {
-                if (i != regexp.length() - 1 && regexp.charAt(i + 1) == '?') {
+            if (!quoted && !escaped && !squareBracketOpen && regexp.charAt(i) == '(') {
```

Comment:
Why don't you have to add a `!quoted` test elsewhere, such as at the test for `')'`?  It would be helpful to write, as a comment, the invariant for the boolean local variables:  what are possible values that they can have at the same time?